internal/s3fs/s3fs.go

```diff
@@ -0,0 +1,257 @@
+package s3fs
+
+import (
+	"context"
+	"errors"
+	"io"
+	"io/fs"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"go.uber.org/zap"
+)
+
+type S3ClientAPI interface {
+	ListObjectsV2(context.Context, *s3.ListObjectsV2Input, ...func(*s3.Options)) (*s3.ListObjectsV2Output, error)
+	GetObject(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+}
+
+// FS is only for accessing files in a single bucket. The directory
+// entries are cached. It is specifically intended for use by a source
+// that calls fs.WalkDir and does not fully implement all fs operations
+type FS struct {
+	logger   *zap.Logger
+	s3Client S3ClientAPI
+
+	bucket   string
+	dirEntry *Dir
+	entries  []fs.DirEntry
+}
+
+// ensure FS implements fs.FS aka Open
+var _ fs.FS = &FS{}
+
+// ensure FS implements fs.StatFS aka Stat
+var _ fs.StatFS = &FS{}
+
+// ensure FS implements fs.ReadDirFS aka ReadDir
+var _ fs.ReadDirFS = &FS{}
+
+// New creates a FS for the single bucket
+func New(logger *zap.Logger, s3Client S3ClientAPI, bucket string) (*FS, error) {
+	return &FS{
+		logger:   logger,
+		s3Client: s3Client,
+		bucket:   bucket,
+	}, nil
+}
+
+// Open implements fs.FS. For the S3 filesystem, it fetches the object
+// contents from s3.
+func (f *FS) Open(name string) (fs.File, error) {
+	if name == "." {
+		return f.dirEntry, nil
+	}
+	pathError := &fs.PathError{
+		Op:   "Open",
+		Path: name,
+		Err:  fs.ErrNotExist,
+	}
+	if !fs.ValidPath(name) {
+		pathError.Err = fs.ErrInvalid
+		return nil, pathError
+	}
+
+	output, err := f.s3Client.GetObject(context.Background(),
+		&s3.GetObjectInput{
+			Bucket: &f.bucket,
+			Key:    &name,
+		})
+	if err != nil {
+		// try to return fs compatible error if possible
+		var nsbe *types.NoSuchBucket
+		if errors.As(err, &nsbe) {
+			return nil, pathError
+		}
+		var nske *types.NoSuchKey
+		if errors.As(err, &nske) {
+			return nil, pathError
+		}
+		var nfe *types.NotFound
+		if errors.As(err, &nfe) {
+			return nil, pathError
+		}
+		pathError.Err = err
+
+		return nil, pathError
+	}
+
+	return &File{
+		bucket:       f.bucket,
+		key:          name,
+		length:       output.ContentLength,
+		body:         output.Body,
+		lastModified: *output.LastModified,
+	}, nil
+}
+
+// Stat implements fs.StatFS. For the s3 filesystem, this gets the
+// objects in the s3 bucket and stores them for later use. Stat can
+// only be called on the currect directory as the s3 filesystem only
+// supports walking a single bucket configured at creation time.
+func (f *FS) Stat(name string) (fs.FileInfo, error) {
```

Comment:
Just for my clarity then: The FS implementation needs to be primed with a call to Stat before it has an entries in it? (unless you happen to know the available keys beforehand of-course.

I see that `fs.WalkDir` starts with a call to `Stat` so that is why that works for the snapshot implementation we use.

I think this is good, I was just contemplating e.g. why we want to perform the operation on Stat and not e.g. on ReadDir instead. I think based on how we interact with `fs.FS` in the snapshot what you have here works great.

Reply:
I was trying to do a minimum viable s3fs and part of that was looking into how `fs.WalkDir` works and seeing that the call to `Stat` will happen before the call to `ReadDir`.

Re-reading the s3fs code, I suppose I could have `ReadDir` populate the filesystem entries if necessary, but because this internal flipt s3fs is only used by `fs.WalkDIr` I didn't think it was necessary or maybe even desirable, as I haven't tested ensuring all the operations work independently of each other.

Said another way, this isn't a general purpose s3 filesystem wrapper and I was trying to be explicit about that.

I'd vote leaving it as it is for now, and if we come up with another use case where this restriction is a problem, we can revisit it then.

Thoughts?